### DISCUSSION TITLE
Understanding GumTree's output

### DISCUSSION THREAD

**Discussion by sunbeomso** [GumTreeDiff/gumtree#292] [Q&A] [answered]
Hello, first of all, thanks for sharing the nice and interesting tool.

I had a question while I am trying out GumTree tool.

Given src.java and dst.java below, GumTree says (I interpreted the result by webdiff command):
i) the method invocation "add(...)" has been "deleted",
ii) "x" , "a" has been "moved"
iii) "+" has been "inserted"

I wondered why GumTree did not use the edit script that uses "update" operation.
In this example, if the edit script is ``replacing add(x,a) by x+a``, wouldn't that be a simpler explanation?

src.java
```
public class Test {
  public int add(int a, int b) {
    return a+b;
  }

  public void test () {
    int x = add(x,a);
  }
}
```

dst.java
```
public class Test {                                                        
  public int add(int a, int b) {                                           
    return a+b;                                                            
  }                                                                        
                                                                           
  public void test () {                                                    
    int x = x + a;
  }                                                                        
}
```

**Comment by jrfaller** [comment, marked as answer]
Hi!  This is currently impossible because add() is of type method call and a + b arithmetic expression.

**Reply by sunbeomso** [reply]
Thanks for the reply!

Can I ask why is it impossible? Specifically,
1) Is "update" operation only valid between nodes of the same type? (e.g., between method calls, between infix expressions)
2) If the answer for the question 1) is "yes", why is such a restriction required? I think "update" can be considered to be valid in some cases (e.g., the example above) even if the types between nodes are different.

**Reply by jrfaller** [reply]
It's exactly that. Moreover update operations only apply to labeled leaf nodes and not on inner nodes such as a binexp or a method call. This is because update is just changing the label of a node, and labels are on leaf nodes in GumTree's AST model.

**Reply by sunbeomso** [reply]
Thanks for the clear answer!
